### PR TITLE
cfengine.py: Updated from upstream

### PR DESCRIPTION
**Pull request by olehermanse**
Copied from:
https://github.com/cfengine/modules/blob/master/libraries/python/cfengine.py

See:
https://github.com/cfengine/modules/pull/6
https://github.com/cfengine/modules/pull/7

### CODE REVIEW

**Comment by olehermanse** [comment]
@cf-bottom jenkins, please

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6676/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6676/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6676/

**Comment by olehermanse** [comment]
@cf-bottom jenkins, please

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6681/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6681/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6681/

**Comment by olehermanse** [comment]
@cf-bottom jenkins, please

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6683/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6683/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6683/

File: misc/custom_promise_types/cfengine.py [hunk 1]
```diff
@@ -161,13 +208,78 @@ def _add_traceback_to_response(self):
...
+    @property
+    def _has_validation_attributes(self):
+        return bool(self._validator_attributes)
+
+    def create_attribute_object(self, promiser, attributes):
+
+        # Check for missing required attributes:
+        for name, attribute in self._validator_attributes.items():
+            if attribute["required"] and name not in attributes:
+                raise ValidationError(f"Missing required attribute '{a}'")
```
Comment (craigcomstock) [inline misc/custom_promise_types/cfengine.py:230]: I don't see an `a` variable defined here for use in the f-string.
Maybe you mean `{name}` instead?

File: misc/custom_promise_types/cfengine.py [hunk 2]
```diff
@@ -161,13 +208,78 @@ def _add_traceback_to_response(self):
...
+        for name in attributes:
+            if name not in self._validator_attributes:
+                raise ValidationError(f"Unknown attribute '{name}'")
+
+        # Check typings and run custom validator callbacks:
+        for name, value in attributes.items():
+            expected = _cfengine_type(self._validator_attributes[name]["typing"])
+            found = _cfengine_type(type(value))
+            if found != expected:
+                raise ValidationError(f"Wrong type for attribute '{name}', requires '{expected}', not '{value}'({found})")
```
Comment (craigcomstock) [inline misc/custom_promise_types/cfengine.py:242]: maybe reword to be more clear?

```
Wrong type for attribute '{name}', requires '{expected}', not '{found}' type as found in '{value}'
```

File: tests/acceptance/30_custom_promise_types/copy_promises.py [hunk 3]
```diff
@@ -6,7 +6,7 @@
 # promise agent copy
 # {
 #     interpreter => "/usr/bin/python3";
-#     path => "$(sys.inputdir)/copy.py";
+#     path => "$(sys.inputdir)/copy_promises.py";
```
Comment (craigcomstock) [inline tests/acceptance/30_custom_promise_types/copy_promises.py:9]: indeed! :+1:

**Review by craigcomstock — COMMENTED** [review]
one typo I think `s/a/name/` in an f-string. Other than that a nit-pick on wording.

File: misc/custom_promise_types/cfengine.py (hunk 1 above)
Comment (olehermanse) [reply misc/custom_promise_types/cfengine.py:230]: Indeed, thanks, I will fix in a follow-up.

File: misc/custom_promise_types/cfengine.py (hunk 2 above)
Comment (olehermanse) [reply misc/custom_promise_types/cfengine.py:242]: I don't think that's more clear.

File: misc/custom_promise_types/cfengine.py (hunk 1 above)
Comment (olehermanse) [reply misc/custom_promise_types/cfengine.py:230]: Addressed as first commit in: https://github.com/cfengine/core/pull/4700
